communities/src/migrations.rs

```diff
@@ -0,0 +1,474 @@
+use super::*;
+use frame_support::{pallet_prelude::*, storage_alias, traits::OnRuntimeUpgrade};
+
+/// The log target.
+const TARGET: &str = "communities::migration::v1";
+
+mod v0 {
+	use super::*;
+
+	#[storage_alias]
+	pub type CommunityIdentifiers<T: Config> =
+		StorageValue<Pallet<T>, Vec<CommunityIdentifier>, ValueQuery>;
+
+	#[storage_alias]
+	pub(super) type CommunityIdentifiersByGeohash<T: Config> =
+		StorageMap<Pallet<T>, Identity, GeoHash, Vec<CommunityIdentifier>, ValueQuery>;
+
+	#[storage_alias]
+	pub(super) type Locations<T: Config> = StorageDoubleMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		CommunityIdentifier,
+		Identity,
+		GeoHash,
+		Vec<Location>,
+		ValueQuery,
+	>;
+
+	#[storage_alias]
+	pub(super) type Bootstrappers<T: Config> = StorageMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		CommunityIdentifier,
+		Vec<<T as frame_system::Config>::AccountId>,
+		ValueQuery,
+	>;
+
+	#[storage_alias]
+	pub(super) type CommunityMetadata<T: Config> = StorageMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		CommunityIdentifier,
+		CommunityMetadataType,
+		ValueQuery,
+	>;
+}
+
+pub mod v1 {
+	use super::*;
+
+	pub struct Migration<T>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config + frame_system::Config> OnRuntimeUpgrade for Migration<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			assert_eq!(StorageVersion::get::<Pallet<T>>(), 0, "can only upgrade from version 0");
+
+			let cid_count = v0::CommunityIdentifiers::<T>::get().len() as u32;
+			log::info!(target: TARGET, "{} cids will be migrated.", cid_count,);
+			ensure!(cid_count <= T::MaxCommunityIdentifiers::get(), "too many cids");
+
+			let cids_by_geohash = v0::CommunityIdentifiersByGeohash::<T>::iter();
+			let mut cids_by_geohash_count = 0u32;
+			for cids in cids_by_geohash {
+				let count = cids.1.len() as u32;
+				ensure!(
+					count <= T::MaxCommunityIdentifiersPerGeohash::get(),
+					"too many cids per geohash"
+				);
```

Comment:
What happens if the `pre_upgrade` checks fails?

Reply:
in production nothing. `pre_upgrade` and `post_upgrade` are just for testing.